analyzers/src/SonarAnalyzer.CSharp/Rules/NamedPlaceholdersShouldBeUnique.cs

```diff
@@ -0,0 +1,122 @@
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+namespace SonarAnalyzer.Rules.CSharp;
+
+[DiagnosticAnalyzer(LanguageNames.CSharp)]
+public sealed class NamedPlaceholdersShouldBeUnique : SonarDiagnosticAnalyzer
+{
+    private const string DiagnosticId = "S6677";
+    private const string MessageFormat = "Message template placeholder '{0}' is not unique.";
+
+    private static readonly HashSet<string> MicrosoftExtensionsLoggingMethods =
+        [
+            "Log",
+            "LogCritical",
+            "LogDebug",
+            "LogError",
+            "LogInformation",
+            "LogTrace",
+            "LogWarning"
+        ];
+
+    private static readonly DiagnosticDescriptor Rule = DescriptorFactory.Create(DiagnosticId, MessageFormat);
+
+    public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics =>
+        ImmutableArray.Create(Rule);
+
+    protected override void Initialize(SonarAnalysisContext context) =>
+        context.RegisterNodeAction(c =>
+            {
+                if (c.Node is InvocationExpressionSyntax invocation
+                    && TemplateArgument(invocation, c.SemanticModel) is { } argument
+                    && MessageTemplates.Parse(argument.Expression.ToString()) is { Success: true } result)
+                {
+                    RaiseAtDuplicates(c, argument, result.Placeholders);
+                }
+            },
+            SyntaxKind.InvocationExpression);
+
+    private static void RaiseAtDuplicates(SonarSyntaxNodeReportingContext context, ArgumentSyntax templateArgument, MessageTemplates.Placeholder[] placeholders)
+    {
+        var duplicates = placeholders
+            .Where(x => !int.TryParse(x.Name, out var _)) // exclude index placeholders like {42}
+            .GroupBy(x => x.Name)
+            .Where(x => x.Count() > 1).Select(x => x.Skip(1))
+            .SelectMany(x => x);
+
+        foreach (var placeholder in duplicates)
+        {
+            var spanStart = templateArgument.Expression.GetLocation().SourceSpan.Start + placeholder.Start;
+            var location = Location.Create(context.Tree, new(spanStart, placeholder.Length));
+            context.ReportIssue(Diagnostic.Create(Rule, location, placeholder.Name));
```

Comment:
Here I had a choice:
- Raise on the entire template
- Raise on each placeholder

I think it's nicer to raise separate, distinct issues on every duplicate placeholder.
I am open to change this if there are disagreements.